modules/11_subclustering/subclustering.qmd

```diff
@@ -0,0 +1,676 @@
+source("R/functions_io.R")
+source("R/functions_plotting.R")
+source("R/functions_analysis.R")
+
+# Be verbose or not
+verbose = param("verbose")
+
+# Parallelisation plan for all functions that support future
+plan(multisession, workers=param("cores"))
+
+# DIRECTORIES
+# Module directory 'results' contains all output that should be provided together with the final report of the project
+dir.create(file.path(module_dir, "results"), showWarnings=FALSE)
+files = list.files(path=file.path(module_dir, "results"), full.names=TRUE)
+if (length(files) > 0) unlink(files, recursive=TRUE)
+
+# Module directory 'sc' contains the final Seurat object
+dir.create(file.path(module_dir, "sc"), showWarnings=FALSE)
+files = list.files(path=file.path(module_dir, "sc"), full.names=TRUE)
+if (length(files) > 0) unlink(files, recursive=TRUE)
+
+# SEURAT OBJECT
+# Read in seurat object
+if (!is.null(param("prev_module_dir"))) {
+  prev_module_dir = param("prev_module_dir")
+} else {
+  prev_module_dir = PreviousModuleDir(module_dir)
+}
+prev_sc_obj = file.path(prev_module_dir, "sc", "sc.rds")
+if(!file.exists(prev_sc_obj)) {
+  stop(FormatString("Could not find a sc.rds file in {prev_module_dir}. Was the respective module already run?"))
+}
+sc = SeuratObject::LoadSeuratRds(prev_sc_obj)
+
+# Set default assay standard
+default_assay = param("default_assay")
+if (is.null(default_assay)) default_assay = Seurat::DefaultAssay(sc)
+Seurat::DefaultAssay(sc) = default_assay
+
+# ON-DISK LAYERS (if used)
+on_disk_counts = param("on_disk_counts")
+
+# If requested, move on-disk layers to faster temp location for faster access
+#on_disk_use_tmp = param("on_disk_use_tmp")
+
+# Note: counts data is not used in this module, therefore deactivated
+on_disk_use_tmp = FALSE
+if (on_disk_counts & on_disk_use_tmp) {
+  on_disk_path = tempdir()
+  with_progress({
+    sc = UpdateMatrixDirs(sc, dir=on_disk_path)
+  }, enable=verbose)
+}
+
+# Variable to indicate if barcodes were removed from the analysis
+# In this case, when using on-disk matrices, they need to be written to disk again
+barcodes_removed = FALSE
+```
+
+<!-- 
+## Remove cells
+
+USER_INPUT:
+
+- Remove the HTML comments
+- Set eval in the chunk options to true
+- Add explanation why barcodes are remove here
+
+```{r}
+#| label: subclustering_remove_barcodes_1
+#| results: asis
+#| eval: false
+
+# Use this chunk to remove barcodes from the analysis.
+# However note that it might be better to remove these barcodes already in the QC module.
+
+# USER INPUT: 
+# - Specify which barcodes of the default assay to remove
+bcs = SeuratObject::Cells(sc[[default_assay]])
+barcode_metadata = sc[[]][bcs, ]
+
+barcodes_to_remove = barcode_metadata %>%
+  dplyr::filter(seurat_clusters == "1") %>%
+  rownames()
+
+# Filter barcodes:
+# a) barcodes that are present only in this assay, will be removed from the whole Seurat object
+# b) barcodes that are present in other assays, will be removed from this assay only
```

Comment:
in other assays and not this one, right?

Reply:
I removed the chunk for filtering barcodes.

But yes barcodes that are filtered in this assay may still be useful in other assays. That is why they are only removed if they are filtered in all assays.